internal/pubsub/subscriber.go

```diff
@@ -0,0 +1,372 @@
+	}
+
+	js, err := c.nc.JetStream()
+	if err != nil {
+		return nil, err
+	}
+
+	c.js = js
+	c.stream = cfg.Stream
+	c.consumer = cfg.Consumer
+	c.prefix = cfg.Prefix
+
+	return &c, nil
+}
+
+func (c *Client) ensureStream() error {
+	c.logger.Debugw("checking that NATS stream exists", "stream_name", c.stream)
+
+	_, err := c.js.StreamInfo(c.stream)
+	if err == nil {
+		c.logger.Debugw("stream exists, not recreating", "stream_name", c.stream)
+		return nil
+	}
+
+	if !errors.Is(err, nats.ErrStreamNotFound) {
+		return err
+	}
+
+	_, err = c.js.AddStream(&nats.StreamConfig{
+		Name: c.stream,
+		Subjects: []string{
+			c.prefix + ".>",
+		},
+		Storage:   nats.FileStorage,
+		Retention: nats.LimitsPolicy,
+		Discard:   nats.DiscardNew,
+	})
+
+	return err
+}
+
+// Listen ensures a stream exists, binds to it, and listens for resource lifecycle events on that stream.
+func (c *Client) Listen() error {
+	// Ensure stream exists before we attempt to listen on it
+	err := c.ensureStream()
+	if err != nil {
+		return err
+	}
+
+	// Set subscription options. We specifically want manual acks in case something goes wrong
+	// persisting the event
+	subOpts := []nats.SubOpt{
+		nats.BindStream(c.stream),
+		nats.Durable(c.consumer),
+		nats.ManualAck(),
+		nats.AckExplicit(),
+	}
+
+	for _, name := range c.resourceTypeNames {
+		subject := fmt.Sprintf("%s.%s.>", c.prefix, name)
+		queueName := fmt.Sprintf("%s-%s", c.consumer, name)
+
+		subscription, err := c.js.QueueSubscribe(subject, queueName, c.receiveMsg, subOpts...)
+		if err != nil {
+			return err
+		}
+
+		c.subscriptions = append(c.subscriptions, subscription)
+	}
+
+	return nil
+}
+
+// Stop drains all subscriptions for the client.
+func (c *Client) Stop() error {
+	for _, sub := range c.subscriptions {
+		err := sub.Drain()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) ackWithError(msg *nats.Msg, err error) {
+	c.logger.Warnw("invalid message - will not reprocess", "error", err.Error())
+
+	err = msg.Ack()
+	if err != nil {
+		c.logger.Errorw("error acking message", "error", err.Error())
+	}
+}
+
+func (c *Client) newResourceFromString(urnStr string) (types.Resource, error) {
+	urn, err := urnx.Parse(urnStr)
+	if err != nil {
+		return types.Resource{}, err
+	}
+
+	return c.qe.NewResourceFromURN(urn)
+}
+
+func (c *Client) createRelationships(ctx context.Context, msg *nats.Msg, resource types.Resource, fields map[string]string) error {
+	var relationships []types.Relationship
+
+	// Attempt to create relationships from the message fields. If this fails, reject the message
+	for field, value := range fields {
+		relation, found := strings.CutSuffix(field, fieldRelationshipSuffix)
+		if !found {
+			continue
+		}
+
+		subjResource, err := c.newResourceFromString(value)
+		if err != nil {
+			c.logger.Errorw("error parsing field - will not reprocess", "event_type", eventTypeCreate, "field", field, "error", err.Error())
+			return msg.Ack()
```

Comment:
all of these buried `Acks` and `Nacks` make me a little uncomfortable - I wonder if there's a better way to handle that?  I realize *why* its this way - ie. how to handle a bad message is different in each case, just a thought.

Reply:
Addressed this by changing the semantics up a bit - we only `Nak` now if we get an error handling the event and don't consider malformed messages to be errors.